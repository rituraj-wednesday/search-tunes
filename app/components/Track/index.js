/**
 *
 * Track
 *
 */

import React, { memo, useEffect } from 'react';
import PropTypes from 'prop-types';
import styled from '@emotion/styled';
import IconButton from '@mui/material/IconButton';
import PlayCircleFilledRoundedIcon from '@mui/icons-material/PlayCircleFilledRounded';
import PauseCircleFilledRoundedIcon from '@mui/icons-material/PauseCircleFilledRounded';
import InfoRoundedIcon from '@mui/icons-material/InfoRounded';
import { If } from '../If/index';
import audioController from '@app/utils/audioController';
import { translate } from '@app/utils/index';

const TrackArtHoverLayer = styled.div`
  height: 100%;
  width: 100%;
  background: #000000b0;
  position: relative;
  top: ${(props) => (props.isPlaying ? '0px' : 'calc(100% - 46px)')};
  transition: top 200ms ease;
`;

const TrackWrapper = styled.div`
  border-radius: 4px;
  border: 3px solid #a88d8d;
  overflow: hidden;
  height: 200px;
  width: 200px;
  background-image: url(${(props) => props.trackArtURL});
  background-size: cover;
  :hover {
    ${TrackArtHoverLayer} {
      top: 0;
    }
  }
`;

const InfoIconWrapper = styled.div`
  width: 50%;
  height: calc(100% - 48px);
  display: flex;
`;

const ButtonFlex = styled.div`
  display: flex;
  height: 100%;
`;

const TitleWrapper = styled.div`
  color: white;
  font-size: 24px;
  height: 46px;
  font-weight: 500;
  padding: 6px;
  width: 100%;
  overflow: hidden;
  white-space: nowrap;
  text-overflow: ellipsis;
  text-align: center;
`;

/**
 * Track component displays track as a tile. Also displays Preview and Info button on hover.
 *
 * @date 09/05/2024 - 16:37:00
 *
 * @param {Object} props - The component props.
 * @param {string} props.track - Track data.
 * @returns {JSX.Element} The Track component.
 */
export function Track({ track, currentTrackID }) {
  const { trackId } = track;
  const isPlaying = currentTrackID === trackId;

  useEffect(() => {
    audioController.registerAudio(track);
  }, [track.trackId]);

  const onPlayPauseClick = () => {
    if (!isPlaying) {
      audioController.play(track);
    } else {
      audioController.stop();
    }
  };

  return (
    <TrackWrapper title={track.trackName} trackArtURL={track.artworkUrl100} aria-label={track.trackName}>
      <TrackArtHoverLayer isPlaying={isPlaying}>
        <TitleWrapper>{track.trackName}</TitleWrapper>
        <ButtonFlex>
          <InfoIconWrapper>
            <IconButton
              sx={{
                margin: 'calc(30% - 4px) 8px auto auto',
                height: '60px',
                width: '60px'
              }}
              aria-label={`${isPlaying ? translate('pauseText') : translate('playText')} ${translate('buttonText')}`}
              onClick={onPlayPauseClick}
            >
              <If
                condition={!isPlaying}
                otherwise={<PauseCircleFilledRoundedIcon sx={{ color: 'white', fontSize: '60px' }} />}
              >
                <PlayCircleFilledRoundedIcon sx={{ color: 'white', fontSize: '60px' }} />
              </If>
            </IconButton>
          </InfoIconWrapper>
          <InfoIconWrapper>
            <IconButton
              sx={{
                margin: 'calc(30% - 4px) auto auto 8px',
                height: '60px',
                width: '60px'
              }}
            >
              <InfoRoundedIcon sx={{ color: 'white', fontSize: '60px' }} />
            </IconButton>
          </InfoIconWrapper>
        </ButtonFlex>
      </TrackArtHoverLayer>
    </TrackWrapper>
  );
}

Track.propTypes = {
  track: PropTypes.shape({
    artworkUrl100: PropTypes.string,
    trackName: PropTypes.string,
    trackId: PropTypes.number,
    previewUrl: PropTypes.string
  }),
  currentTrackID: PropTypes.number
};

Track.defaultProps = {
  track: {
    artworkUrl100: '',
    trackName: ''
  }
};

export default memo(Track);
