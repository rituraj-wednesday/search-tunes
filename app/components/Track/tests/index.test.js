/**
 *
 * Tests for Track
 *
 */

import React from 'react';
// import { fireEvent } from '@testing-library/dom'
import { renderWithIntl } from '@utils/testUtils';
import Track from '../index';

describe('<Track />', () => {
  it('should render and match the snapshot', () => {
    const { baseElement } = renderWithIntl(<Track track={{ trackName: 'hello', artworkUrl100: 'artWork' }}/>);
    expect(baseElement).toMatchSnapshot();
  });

  it('should contain 1 Track component', () => {
    const { getAllByTestId } = renderWithIntl(<Track />);
    expect(getAllByTestId('tune-tile').length).toBe(1);
  });
});
