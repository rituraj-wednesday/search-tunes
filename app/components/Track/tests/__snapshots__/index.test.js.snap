// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Track /> should render and match the snapshot 1`] = `
<body>
  <div>
    <div
      aria-label="hello"
      class="css-1exn88k e1b9j16n1"
      data-testid="tune-tile"
    >
      <div
        class="css-1i846qt e1b9j16n2"
      >
        <div
          class="css-d9tof8 e1b9j16n0"
        >
          <button
            class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-1wcb6rg-MuiButtonBase-root-MuiIconButton-root"
            tabindex="0"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-11bbgqr-MuiSvgIcon-root"
              data-testid="PlayCircleFilledRoundedIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2m-2 13.5v-7c0-.41.47-.65.8-.4l4.67 3.5c.27.2.27.6 0 .8l-4.67 3.5c-.33.25-.8.01-.8-.4"
              />
            </svg>
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </div>
        <div
          class="css-d9tof8 e1b9j16n0"
        >
          <button
            class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-1tlcx8h-MuiButtonBase-root-MuiIconButton-root"
            tabindex="0"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-11bbgqr-MuiSvgIcon-root"
              data-testid="InfoRoundedIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2m0 15c-.55 0-1-.45-1-1v-4c0-.55.45-1 1-1s1 .45 1 1v4c0 .55-.45 1-1 1m1-8h-2V7h2z"
              />
            </svg>
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </div>
      </div>
    </div>
  </div>
</body>
`;
